File: coding-interview-patterns/scala/src/main/scala/patterns/Stacks.scala
package patterns
object Stacks {
  def nextGreater(nums:Array[Int]): Array[Int] = {
    val st = scala.collection.mutable.Stack[Int]()
    val res = Array.fill(nums.length)(-1)
    for (i<-nums.indices){
      while (st.nonEmpty && nums(st.top) < nums(i)) res(st.pop()) = nums(i)
      st.push(i)
    }
    res
  }
  def dailyTemperatures(T:Array[Int]): Array[Int] = {
    val st = scala.collection.mutable.Stack[Int]()
    val res = Array.fill(T.length)(0)
    for (i<-T.indices){
      while (st.nonEmpty && T(st.top) < T(i)){
        val j=st.pop(); res(j)=i-j
      }
      st.push(i)
    }
    res
  }
  def largestRectangle(h:Array[Int]): Int = {
    val st=scala.collection.mutable.Stack[Int]()
    val a=h :+ 0
    var ans=0
    for (i<-a.indices){
      while (st.nonEmpty && a(st.top) > a(i)){
        val H=a(st.pop()); val L= if (st.isEmpty) -1 else st.top
        ans = ans.max(H * (i - L - 1))
      }
      st.push(i)
    }
    ans
  }
}

File: coding-interview-patterns/python/patterns/stacks.py
def next_greater(nums):
    st=[]; res=[-1]*len(nums)
    for i,x in enumerate(nums):
        while st and nums[st[-1]]<x:
            res[st.pop()]=x
        st.append(i)
    return res

def daily_temperatures(T):
    st=[]; res=[0]*len(T)
    for i,t in enumerate(T):
        while st and T[st[-1]]<t:
            j=st.pop(); res[j]=i-j
        st.append(i)
    return res

def largest_rectangle(h):
    st=[]; ans=0; h.append(0)
    for i,x in enumerate(h):
        while st and h[st[-1]]>x:
            H=h[st.pop()]; L=st[-1] if st else -1
            ans=max(ans, H*(i-L-1))
        st.append(i)
    h.pop(); return ans
